src/webgpu/api/operation/command_buffer/queries/occlusionQuery.spec.ts

```diff
@@ -0,0 +1,1076 @@
+        stencilLoadOp: 'clear',
+        stencilStoreOp: 'store',
+      },
+      occlusionQuerySet,
+    };
+
+    await t.runQueryTest(
+      resources,
+      renderPassDescriptor,
+      (helper, queryIndex) => {
+        helper.setStencilReference(queryIndex % 2);
+        const queryHelper = helper.beginOcclusionQuery(queryIndex);
+        queryHelper.setPipeline(pipeline);
+        queryHelper.setVertexBuffer(vertexBuffer);
+        queryHelper.draw(3);
+        queryHelper.end();
+      },
+      (passed, queryIndex) => {
+        const expectPassed = queryIndex % 2 === 0;
+        t.expect(
+          !!passed === expectPassed,
+          `queryIndex: ${queryIndex}, was: ${!!passed}, expected: ${expectPassed}, ${name}`
+        );
+      }
+    );
+  });
+
+g.test('occlusion_query,sample_mask')
+  .desc(
+    `
+      Test beginOcclusionQuery/endOcclusionQuery using sample_mask to occlude
+
+      Set mask to 4 and draw quads in opposing corners of the texel.
+    `
+  )
+  .params(kQueryTestBaseParams.combine('sampleMask', [0, 2, 4, 6]))
+  .fn(async t => {
+    const { writeMask, renderMode, bufferOffset, querySetOffset, sampleMask } = t.params;
+    const kNumQueries = 30;
+    const sampleCount = 4;
+    const resources = t.setup({
+      writeMask,
+      renderMode,
+      bufferOffset,
+      querySetOffset,
+      numQueries: kNumQueries,
+      sampleCount,
+      sampleMask,
+    });
+    const { occlusionQuerySet, renderTargetTexture, pipeline } = resources;
+
+    const createQuad = (offset: number) => {
+      // prettier-ignore
+      return t.createVertexBuffer(new Float32Array([
+        offset + 0   , offset + 0   , 0,
+        offset + 0.25, offset + 0   , 0,
+        offset + 0   , offset + 0.25, 0,
+        offset + 0   , offset + 0.25, 0,
+        offset + 0.25, offset + 0   , 0,
+        offset + 0.25, offset + 0.25, 0,
+      ]));
+    };
+
+    const vertexBufferTL = createQuad(0);
+    const vertexBufferBR = createQuad(0.25);
+
+    const multisampleRenderTarget = t.createTexture({
+      size: kTextureSize,
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.RENDER_ATTACHMENT,
+      sampleCount,
+    });
+
+    const renderPassDescriptor: GPURenderPassDescriptor = {
+      colorAttachments: [
+        {
+          view: multisampleRenderTarget.createView(),
+          resolveTarget: renderTargetTexture.createView(),
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+      occlusionQuerySet,
+    };
+
+    await t.runQueryTest(
+      resources,
+      renderPassDescriptor,
+      (helper, queryIndex) => {
+        const queryHelper = helper.beginOcclusionQuery(queryIndex);
+        queryHelper.setPipeline(pipeline);
+        queryHelper.setVertexBuffer(queryIndex % 2 ? vertexBufferBR : vertexBufferTL);
+        queryHelper.draw(6);
+        queryHelper.end();
+      },
+      (passed, queryIndex) => {
+        const drawMask = queryIndex % 2 ? 2 : 4;
```

Comment:
Sorry, I read this over a couple of times and still wasn't able to understand this condition... Could you explain it a bit?

Reply:
added a comment. Is it ok?

Reply:
Ah! Turns out the misunderstanding was due to my lack of understanding of how the sample mask worked (and how the samples are indexed.) It all makes sense now! Kai pointed me to https://learn.microsoft.com/en-us/windows/win32/api/d3d11/ne-d3d11-d3d11_standard_multisample_quality_levels which helped me understand it better. Optional if you want to include the link in the comment too to help any future confused readers like me 😅